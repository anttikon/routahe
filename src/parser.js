import moment from 'moment'

const timeParseFormats = ['HH:mm', 'H:mm']
const dateParseFormats = [
  'DD.MM.YYYY',
  'DD.MM.',
  'DD.MM',
  'DD.M.YYYY',
  'DD.M.',
  'DD.M',
  'D.M',
  'D.MM',
  'D.MM.',
  'D.M.YYYY',
  'D.MM.YYYY',
]

function parseTime(time) {
  if (time && time.startsWith('@')) {
    return moment(time.slice(1), timeParseFormats, true)
  } else {
    return moment(time, timeParseFormats, true)
  }
}

export const parseDateTime = opts => {
  const parsedDate = moment(opts.date, dateParseFormats, true)
  const parsedTime = parseTime(opts.time)

  const date = parsedDate.isValid() ? parsedDate : new moment()
  const time = parsedTime.isValid() ? parsedTime : new moment()

  time.set({ date: date.date(), month: date.month(), year: date.year() })

  return time
}

export const isTime = arg => arg && !!arg.match(/^@?\d{1,2}:\d{2}$/g)

export const isDate = arg => arg && !!arg.match(/^\d{1,2}\.\d{1,2}\.?(\d{4})?$/g)

const filterUndefined = array => array.filter(obj => !!obj)

export const parseArguments = args => {
  const definedArgs = filterUndefined(args)

  const mapped = definedArgs.reduce((opts, arg) => {
    if (isTime(arg)) {
      opts.time = arg
    } else if (isDate(arg)) {
      opts.date = arg
    } else if (!opts.from) {
      opts.from = arg
    } else if (!opts.to) {
      opts.to = arg
    }

    if (isTime(arg) && arg.startsWith('@')) {
      opts.arriveBy = true
    }

    return opts
  }, {})

  mapped.dateTime = parseDateTime(mapped)

  return mapped
}
