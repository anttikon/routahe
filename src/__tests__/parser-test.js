import { parseArguments, isDate, isTime } from '../parser'

const from = 'Tehtaankatu 24'
const to = 'Koirapuisto'
const date = '8.8.'
const time = '21:00'

describe('parseArguments', () => {
  it('should parse arguments with format: from to date time', () => {
    expect(parseArguments([from, to, date, time])).toMatchSnapshot()
  })

  it('should parse arguments with format: from to time date', () => {
    expect(parseArguments([from, to, time, date])).toMatchSnapshot()
  })

  it('should parse arguments with format: date time from to', () => {
    expect(parseArguments([date, time, from, to])).toMatchSnapshot()
  })

  it('should parse arguments with format: time date from to', () => {
    expect(parseArguments([time, date, from, to])).toMatchSnapshot()
  })
})

describe('isDate', () => {
  it('should return true with 1.1', () => {
    expect(isDate('1.1')).toBeTruthy()
  })

  it('should return true with 1.1.', () => {
    expect(isDate('1.1.')).toBeTruthy()
  })
  it('should return true with 1.1.2019', () => {
    expect(isDate('1.1.2019')).toBeTruthy()
  })
  it('should return false with 1.', () => {
    expect(isDate('1.')).toBeFalsy()
  })
  it('should return false with 1', () => {
    expect(isDate('1')).toBeFalsy()
  })
  it('should return false with asdf', () => {
    expect(isDate('asdf')).toBeFalsy()
  })
  it('should return false with undefined', () => {
    expect(isDate()).toBeFalsy()
  })
})

describe('isTime', () => {
  it('should return true with 00:00', () => {
    expect(isTime('00:00')).toBeTruthy()
  })
  it('should return true with 0:00', () => {
    expect(isTime('0:00')).toBeTruthy()
  })
  it('should return true with 23:59', () => {
    expect(isTime('23:59')).toBeTruthy()
  })
  it('should return false with 23:59:', () => {
    expect(isTime('23:59:')).toBeFalsy()
  })
  it('should return false with asdf', () => {
    expect(isTime('asdf')).toBeFalsy()
  })
  it('should return false with undefined', () => {
    expect(isTime()).toBeFalsy()
  })
})
