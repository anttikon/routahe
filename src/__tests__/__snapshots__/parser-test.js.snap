// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseArguments should parse arguments with format: date time from to 1`] = `
Object {
  "date": "8.8.",
  "dateTime": "2019-04-23T18:00:00.000Z",
  "from": "Tehtaankatu 24",
  "time": "21:00",
  "to": "Koirapuisto",
}
`;

exports[`parseArguments should parse arguments with format: from to date time 1`] = `
Object {
  "date": "8.8.",
  "dateTime": "2019-04-23T18:00:00.000Z",
  "from": "Tehtaankatu 24",
  "time": "21:00",
  "to": "Koirapuisto",
}
`;

exports[`parseArguments should parse arguments with format: from to time date 1`] = `
Object {
  "date": "8.8.",
  "dateTime": "2019-04-23T18:00:00.000Z",
  "from": "Tehtaankatu 24",
  "time": "21:00",
  "to": "Koirapuisto",
}
`;

exports[`parseArguments should parse arguments with format: time date from to 1`] = `
Object {
  "date": "8.8.",
  "dateTime": "2019-04-23T18:00:00.000Z",
  "from": "Tehtaankatu 24",
  "time": "21:00",
  "to": "Koirapuisto",
}
`;
